xen: finish moving to `by-name`; refactor

SigmaSquadron opened the pull request:
Follow-up to the second Xen refactor:
- #345192

This finally moves the builder to `by-name`, as we can now easily override Xen for a custom build thanks to `finalAttrs`.
Also does some minor refactoring, like enabling `strictDeps` and using `getExe`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron committed:
xen: move main package builder to by-name

This causes the removal of buildXenPackage, as the recommended way to
build a custom Xen is to override the package.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: remove now-unused secondary function header

The old `buildXenPackage` builder was a curried function that had to be
completed with a second call with attributes like `pname` and `version`.
As it is no longer necessary to call a generic builder to build custom
variants of Xen, the package must comply with the `by-name` checks, which
include the requirement that all `package.nix` files in this directory
hierarchy resolve to a derivation. (and not a function)

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: enable strictDeps

Shuffles the input lists to ensure Xen works with `strictDeps` enabled.
There are more derivations in `nativeBuildInputs` than strictly
necessary for a successful build, but Xen calls for the OCaml and dev86
binaries during certain situations when building certain components.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: minor refactoring

Changes some (long overdue) TODO comments and uses `lib.getExe` instead
of hardcoding /bin paths.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

drupol commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,22 +1,388 @@
+      zlib
+      zstd
+    ]
+    ++ optional withFlask checkpolicy
+    ++ optional (versionOlder finalAttrs.version "4.19") systemdMinimal;
```
Minor suggestion:

```suggestion
    ++ optionals withFlask [
      checkpolicy
    ] ++ optionals (versionOlder finalAttrs.version "4.19") [
    systemdMinimal
    ];
```

drupol commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,22 +1,388 @@
+    ];
+
+    teams = [ teams.xen ];
+    knownVulnerabilities = optional (versionOlder finalAttrs.version minSupportedVersion) "The Xen Project Hypervisor version ${finalAttrs.version} is no longer supported by the Xen Project Security Team. See https://xenbits.xenproject.org/docs/unstable/support-matrix.html";
```
```suggestion
    knownVulnerabilities = optionals (versionOlder finalAttrs.version minSupportedVersion) [
    "The Xen Project Hypervisor version ${finalAttrs.version} is no longer supported by the Xen Project Security Team. See https://xenbits.xenproject.org/docs/unstable/support-matrix.html"
    ];
```

SigmaSquadron replied:
I personally prefer `optional` as it saves a couple of keystrokes, but I don't really have that strong of an opinion on it. If it makes things easier for newcomers, I'm all for it.

CertainLach replied:
I would argue that it is easy to understand what the difference between `optional` and `optionals`, and for nixpkgs it is a useful knowledge, as everything else uses `optional` when singleton array is wanted

drupol reviewed (commented):
Added a minor opinonated feedback to switch from `optional` to the more explicit `optionals` (singular to plural form) to reduce the cognitive load for newcomers.

IMO, it also makes sense to me to use it when dealing with lists (plural form), like `buildInputs` or `knownVulnerabilities`.

CertainLach commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,22 +1,389 @@
+  src = fetchgit {
+    url = "https://xenbits.xenproject.org/git-http/xen.git";
+    tag = "RELEASE-${finalAttrs.version}";
+    hash = "sha256-s0eCBCd6ybl+kLtXCC6E1sk++w7txXn/B/Cg5acQFfY=";
```
Shouldn't `hash` also be moved to attrs?

We can't override version without needing to override src, so using `finalAttrs.version` in tag feels wrong.

SigmaSquadron replied:
Well, the idea is that you'd override `src`.

SigmaSquadron replied:
Why? That's not the case in any other package.

CertainLach replied:
Pretty sure there was more issues regarding this antipattern: https://github.com/NixOS/nixpkgs/issues/310373

SigmaSquadron replied:
Well, could just use the commit again.

SigmaSquadron replied:
We're back to using the exact revision, so we no longer reuse `version`.

philiptaron reviewed (approved):
It's a pretty steep deprecation, but I trust the Xen project to judge whether that's too fast.

SigmaSquadron commented:
I should clarify that as a breaking change, this will need to wait for the branch-off.

philiptaron reviewed (approved):
Lots of agreement with the TODOs in the derivation.

Deprecation schedule now looks quite sane.

SigmaSquadron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -41,6 +150,263 @@ buildXenPackage.override { inherit python3Packages; } {
+    # a more consistent output for downstreams who override the
+    # vendor attribute above.
+    ''
+      substituteInPlace xen/Makefile \
+        --replace-fail "\$(D)\$(EFI_MOUNTPOINT)/efi/\$(EFI_VENDOR)/\$(T)-\$(XEN_FULLVERSION).efi" \
+                  "\$(D)\$(BOOT_DIR)/\$(T)-\$(XEN_FULLVERSION).efi"
```
Agreed. I'll whip up a patch soon.

SigmaSquadron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -41,6 +150,263 @@ buildXenPackage.override { inherit python3Packages; } {
+    ''
+
+    # Fix shebangs in Xen's various scripts.
+    # TODO: Patch the individual calls instead.
```
Handling with https://github.com/NixOS/nixpkgs/pull/406638#discussion_r2183216876

When I first refactored Xen, I did not know how to use `replaceVars` inside and `.patch` file.

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -41,6 +150,263 @@ buildXenPackage.override { inherit python3Packages; } {
+    runHook preInstall
+
+    mkdir --parents $out $out/share $boot
+    cp -prvd dist/install/nix/store/*/* $out/
```
Yeah, I'm not fussed. Seeing `rm --recursive --force` instead of `rm -rf` then this combo made me laugh.

SigmaSquadron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -41,6 +150,263 @@ buildXenPackage.override { inherit python3Packages; } {
+
+  passthru = {
+    efi = "boot/xen-${finalAttrs.upstreamVersion}.efi";
+    flaskPolicy =
+      if withFlask then
+        warn "This Xen was compiled with FLASK support, but the FLASK file may not match the Xen version number. Please hardcode the path to the FLASK file instead." "boot/xenpolicy-${finalAttrs.upstreamVersion}"
+      else
+        throw "This Xen was compiled without FLASK support.";
```
I guess we can drop the FLASK stuff, again. I did want to get it working at some point.

SigmaSquadron replied:
Yeah I'm not sure I want to drop the flask stuff. If we build without checkpolicy, the flask stuff is still built, but it won't work.
This may be sort of janky but there's not much we can do here.

SigmaSquadron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -41,6 +150,263 @@ buildXenPackage.override { inherit python3Packages; } {
+      addAutoPatchelfSearchPath $out/lib
+      autoPatchelf $out/libexec/xen/bin
+    ''
+    # Flask is particularly hard to disable. Even after
+    # setting the make flags to `n`, it still gets compiled.
+    # If withFlask is disabled, delete the extra binaries.
```
Uhh I think we talked briefly with Andrew Cooper from upstream Xen about this, but I can't remember what came of it.

Since it's still here, my guess is that it's a consequence of Xen's very bespoke build system that is difficult to fix.

SigmaSquadron committed:
xen: use a patch file to replace external script dependencies

This replaces the brittle PATH wrapper and substituteInPlace calls that
ensured the systemd services and the Xen scripts had all the necessary
dependencies. The new patch can replace the calls to external
executables much more effectively.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -41,6 +150,263 @@ buildXenPackage.override { inherit python3Packages; } {
+    # The following patch fixes the call to /bin/mkdir on the
+    # launch_xenstore.sh helper script.
+    + ''
+      substituteInPlace tools/hotplug/Linux/launch-xenstore.in \
+        --replace-fail "/bin/mkdir" "${getExe' coreutils "mkdir"}"
+    ''
+
+    # The following expression fixes the paths called by Xen's systemd
+    # units, so we can use them in the NixOS module.
+    + ''
+      substituteInPlace \
+        tools/hotplug/Linux/systemd/{xen-init-dom0,xen-qemu-dom0-disk-backend,xenconsoled,xendomains,xenstored}.service.in \
+        --replace-fail /bin/grep ${getExe gnugrep}
+      substituteInPlace \
+       tools/hotplug/Linux/systemd/{xen-qemu-dom0-disk-backend,xenconsoled}.service.in \
+        --replace-fail "/bin/mkdir" "${getExe' coreutils "mkdir"}"
```
So... the patch is two thousand lines long.

philiptaron replied:
oh no

philiptaron commented on pkgs/by-name/xe/xen/0001-makefile-efi-output-directory.patch:
```diff
@@ -0,0 +1,42 @@
+Date: Fri, 4 Jul 2025 18:07:01 +0000
+Subject: [PATCH 1/2] xen/Makefile: patch .efi output directory
+
+This is necessary so the build doesn not fail when Xen tries to install
```
```suggestion
This is necessary so the build does not fail when Xen tries to install
```

philiptaron commented on pkgs/by-name/xe/xen/0002-scripts-external-executable-calls.patch:
```diff
@@ -0,0 +1,2572 @@
+ tools/hotplug/Linux/xen-hotplug-common.sh.in  |   8 +-
+ tools/hotplug/Linux/xen-network-common.sh     |  40 +--
+ tools/hotplug/Linux/xendomains.in             | 240 +++++++++---------
+ 33 files changed, 432 insertions(+), 432 deletions(-)
```
That's actually not too bad. We could make the patch smaller by including less context, but I'm actually not fussed.

SigmaSquadron replied:
Using two lines of context now.

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,13 +1,199 @@
+      "kill"
+      "head"
+      "dirname"
+      "echo"
```
remove me: this is a shell builtin in bash

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,13 +1,199 @@
+      "readlink"
+      "rm"
+      "ls"
+      "kill"
```
remove me: this is a shell builtin in bash.

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,13 +1,199 @@
+      "basename"
+      "cat"
+      "cp"
+      "test"
```
remove me: this is a shell builtin

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,13 +1,199 @@
+      "rmmod"
+    ])
+    // (mkTools libnl (
+      map (str: "nl-qdisc-${str}") [
+        "add"
+        "delete"
+        "list"
+      ]
```
this would be clearer and more easy to search for if spelled out

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -36,11 +222,228 @@ buildXenPackage.override { inherit python3Packages; } {
+      zstd
+    ]
+    ++ optionals withFlask [ checkpolicy ]
+    ++ optionals (versionOlder finalAttrs.version "4.19") [ systemd ];
```
Is this correct: new versions drop the systemd dependency?

philiptaron replied:
That's pretty interesting. Makes sense.

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -36,11 +222,228 @@ buildXenPackage.override { inherit python3Packages; } {
+      ocamlPackages.findlib
+      ocamlPackages.ocaml
+    ]
+    ++ (with python3Packages; [
+      python
+      setuptools
+      wrapPython
+    ]);
```
Isn't this `python3.withPackages (ps: [ ps.setuptools ps.wrapPython ])`?

SigmaSquadron replied:
I'm not sure. I don't think so; since `wrapPython` also includes a shell hook.

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -1,13 +1,199 @@
+      mkTools = pkg: tools: genAttrs tools (tool: getExe' pkg tool);
+    in
+    (genAttrs [
+      "CONFIG_DIR"
+      "CONFIG_LEAF_DIR"
+      "LIBEXEC_BIN"
+      "LINUX_BACKEND_MODULES"
+      "XENSTORED"
+      "XEN_LOG_DIR"
+      "XEN_RUN_DIR"
+      "XEN_SCRIPT_DIR"
+      "qemu_xen_path"
+      "qemu_xen_systemd"
+      "sbindir"
```
All of these show up in the patch file but should not trigger an error from replaceVars: that's why they're mapped to `null`.

SigmaSquadron replied:
I'm not sure what's actionable here? `replaceVars` will fail if those are not mapped to `null`.

```
0002-scripts-external-executable-calls.patch> The following look like unsubstituted Nix identifiers that remain in /nix/store/w5mgkyg1m9j8vfq05bnc9vgzd8mckdyr-0002-scripts-external-executable-calls.patch:
0002-scripts-external-executable-calls.patch> @XEN_SCRIPT_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_LEAF_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_LEAF_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_LEAF_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_LEAF_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_LEAF_DIR@
0002-scripts-external-executable-calls.patch> @LIBEXEC_BIN@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @qemu_xen_systemd@
0002-scripts-external-executable-calls.patch> @XEN_LOG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_DIR@
0002-scripts-external-executable-calls.patch> @CONFIG_LEAF_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @XEN_RUN_DIR@
0002-scripts-external-executable-calls.patch> @sbindir@
0002-scripts-external-executable-calls.patch> @LIBEXEC_BIN@
0002-scripts-external-executable-calls.patch> @LIBEXEC_BIN@
0002-scripts-external-executable-calls.patch> @XEN_SCRIPT_DIR@
0002-scripts-external-executable-calls.patch> Use the more precise `substitute` function if this check is in error.

```

philiptaron commented on pkgs/by-name/xe/xen/package.nix:
```diff
@@ -36,11 +222,228 @@ buildXenPackage.override { inherit python3Packages; } {
+      "XEN_WHOAMI=${finalAttrs.pname}"
+      "XEN_DOMAIN=${finalAttrs.vendor}"
+
+      "GIT=${getExe' coreutils "false"}"
+      "WGET=${getExe' coreutils "false"}"
```
Heh.

SigmaSquadron replied:
Xen's build system will want to download fresh sources from the internet. This prevents it from doing so.

philiptaron reviewed (changes requested):
My quick `rg` math says that dropping the three IMO erroneous substitutions of `echo`, `kill`, and `test` (which are already dependency injected through patching the bash shebang at the top) drops 203 mutations in the patch file.

This looks like the right direction to me, though, megapatch though it may be.

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 406638`
Commit: `5c3b43ab65a72f49893817d88edc82a3d53713b9`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 61 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>collectd</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>docker-machine-kvm2</li>
    <li>fdroidserver</li>
    <li>fdroidserver.dist</li>
    <li>gnome-boxes</li>
    <li>libguestfs (python313Packages.guestfs)</li>
    <li>libguestfs.guestfsd (python313Packages.guestfs.guestfsd)</li>
    <li>librenms</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>libvmi</li>
    <li>libvmi.dev</li>
    <li>libvmi.lib</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>multipass</li>
    <li>perl538Packages.SysVirt</li>
    <li>perl538Packages.SysVirt.devdoc</li>
    <li>perl540Packages.SysVirt</li>
    <li>perl540Packages.SysVirt.devdoc</li>
    <li>podman-bootc</li>
    <li>prometheus-libvirt-exporter</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.guestfsd</li>
    <li>python312Packages.libvirt</li>
    <li>python312Packages.libvirt.dist</li>
    <li>python312Packages.xen</li>
    <li>python312Packages.xen.boot</li>
    <li>python312Packages.xen.dev</li>
    <li>python312Packages.xen.doc</li>
    <li>python312Packages.xen.man</li>
    <li>python313Packages.libvirt</li>
    <li>python313Packages.libvirt.dist</li>
    <li>xen (python313Packages.xen)</li>
    <li>xen.boot (python313Packages.xen.boot)</li>
    <li>xen.dev (python313Packages.xen.dev)</li>
    <li>xen.doc (python313Packages.xen.doc)</li>
    <li>xen.man (python313Packages.xen.man)</li>
    <li>qemu_xen</li>
    <li>qemu_xen.debug</li>
    <li>qemu_xen.doc</li>
    <li>qemu_xen.ga</li>
    <li>rubyPackages.ruby-libvirt (rubyPackages_3_3.ruby-libvirt)</li>
    <li>rubyPackages_3_1.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>rubyPackages_3_4.ruby-libvirt</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-top</li>
    <li>virt-v2v</li>
    <li>virt-viewer</li>
    <li>virtnbdbackup</li>
    <li>virtnbdbackup.dist</li>
    <li>xen-guest-agent</li>
  </ul>
</details>
